crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -750,41 +750,58 @@ unsafe impl<'a, T: Resource> ReadOnlySystemParam for Res<'a, T> {}
 // SAFETY: Res ComponentId access is applied to SystemMeta. If this Res
 // conflicts with any prior access, a panic will occur.
 unsafe impl<'a, T: Resource> SystemParam for Res<'a, T> {
-    type State = ComponentId;
+    type State = (ComponentId, ComponentId);
     type Item<'w, 's> = Res<'w, T>;
 
     fn init_state(world: &mut World) -> Self::State {
-        world.components_registrator().register_resource::<T>()
+        (
+            world.components_registrator().register_resource::<T>(),
+            world
+                .components_registrator()
+                .register_component::<IsResource>(),
```

Comment:
It might be worth registering `IsResource` in `World::bootstrap` using a constant `ComponentId` like we do with lifecycle events.  That would avoid needing to store the extra ID in every `Res` parameter.